albatross/evaluation/cross_validation.h

```diff
@@ -0,0 +1,326 @@
+  auto get_predictions(const RegressionDataset<FeatureType> &dataset) const {
+    const auto indexer = leave_one_out_indexer(dataset.features);
+    const auto folds = folds_from_fold_indexer(dataset, indexer);
+    return albatross::get_predictions(model_, folds);
+  }
+
+  // get_prediction
+
+  /*
+   * get_prediction deals with returning a Prediction object which looks
+   * like any other but which provides out of sample predictions for
+   * a given indexer.
+   */
+  template <typename FeatureType>
+  CVPrediction<ModelType, FeatureType>
+  get_prediction(const RegressionDataset<FeatureType> &dataset,
+                 const FoldIndexer &indexer) const {
+    return CVPrediction<ModelType, FeatureType>(model_, dataset, indexer);
+  }
+
+  template <typename FeatureType>
+  auto get_prediction(const RegressionDataset<FeatureType> &dataset,
+                      const GrouperFunction<FeatureType> &grouper) const {
+    const auto indexer = leave_one_group_out_indexer(dataset.features, grouper);
+    return get_prediction(dataset, indexer);
+  }
+
+  template <typename FeatureType>
+  auto get_prediction(const RegressionDataset<FeatureType> &dataset) const {
+    const auto indexer = leave_one_out_indexer(dataset.features);
+    return get_prediction(dataset, indexer);
+  }
+
+  // Scores
+
+  template <typename RequiredPredictType, typename FeatureType>
+  Eigen::VectorXd
+  scores(const EvaluationMetric<RequiredPredictType> &metric,
+         const std::vector<RegressionFold<FeatureType>> &folds) const {
+    const auto preds = get_predictions(folds);
+    std::vector<RequiredPredictType> predictions;
+    for (const auto &pred : preds) {
+      predictions.emplace_back(pred.template get<RequiredPredictType>());
+    }
+    return cross_validated_scores(metric, folds, predictions);
+  }
+
+  template <typename RequiredPredictType, typename FeatureType>
+  Eigen::VectorXd scores(const EvaluationMetric<RequiredPredictType> &metric,
+                         const RegressionDataset<FeatureType> &dataset,
+                         const FoldIndexer &indexer) const {
+    const auto folds = folds_from_fold_indexer(dataset, indexer);
+    const auto prediction = get_prediction(dataset, indexer);
+    const auto predictions =
+        prediction.template get<std::vector<RequiredPredictType>>();
+    return cross_validated_scores(metric, folds, predictions);
+  }
+
+  template <typename RequiredPredictType, typename FeatureType>
+  Eigen::VectorXd scores(const EvaluationMetric<RequiredPredictType> &metric,
+                         const RegressionDataset<FeatureType> &dataset,
+                         const GrouperFunction<FeatureType> &grouper) const {
+    const auto indexer = leave_one_group_out_indexer(dataset.features, grouper);
+    return scores(metric, dataset, indexer);
+  }
+
+  template <typename RequiredPredictType, typename FeatureType>
+  Eigen::VectorXd scores(const EvaluationMetric<RequiredPredictType> &metric,
```

Comment:
yeah, I could see it being confusing.  As an alternative I could just remove these special case functions and force the user to do:
`scores(metric, dataset, leave_one_out_indexer(dataset.features))`.
which would definitely make it more explicit.

Or yet another option I considered was to completely remove the `get_prediction` part:

`model.cross_validate().leave_one_out(dataset).scores()`

Reply:
What do you think about this:
```
  template <typename FeatureType>
  CVPrediction<ModelType, FeatureType>
  leave_out(const RegressionDataset<FeatureType> &dataset,
            const FoldIndexer &indexer) const {
    return CVPrediction<ModelType, FeatureType>(model_, dataset, indexer);
  }

  template <typename FeatureType>
  auto leave_one_group_out(const RegressionDataset<FeatureType> &dataset,
                           const GrouperFunction<FeatureType> &grouper) const {
    const auto indexer = leave_one_group_out_indexer(dataset.features, grouper);
    return leave_out(dataset, indexer);
  }

  template <typename FeatureType>
  auto leave_one_out(const RegressionDataset<FeatureType> &dataset) const {
    const auto indexer = leave_one_out_indexer(dataset.features);
    return leave_out(dataset, indexer);
  }
```
It strays from the syntax of the models but it's more specific for cross validation.
```
Eigen::VectorXd mean = model.get_fit_model(dataset).get_predictions(dataset.features).mean();
Eigen::VectorXd loo_mean = model.cross_validate().leave_one_out(dataset).mean();
```
vs
```
Eigen::VectorXd mean = model.get_fit_model(dataset).get_predictions(dataset.features).mean();
Eigen::VectorXd loo_mean = model.cross_validate().get_predictions(dataset).mean();
```

Reply:
Personally, I'm just looking for the fold indices and metric to be explicit. Being able to return the prediction and then compute scores is good, as it allows multiple scores to be computed, and also allows for refolding - as we do now. This latest proposal looks good to me.
But `scores(metric, dataset, leave_one_out_indexer(dataset.features))` hides this.

Reply:
Yeah, Re:computing multiple metrics I'm trying to set this all up in a way where if you'd rather you can do everything manually using free functions the way it was before this change.  Ie,

```
auto folds = folds_from_fold_indexer(dataset, indexer);
auto predictions = get_predictions(model, folds);
auto foo_scores = cross_validated_scores(foo_metric, folds, predictions);
auto bar_scores = cross_validated_scores(bar_metric, folds, predictions);
```

Re:naming I'm kind of torn about being super explicit about each variant of computing folds `leave_one_out_scores`, `leave_one_group_out_scores` instead of just having a unified `scores` method whose behavior is (hopefully) obvious from the arguments.  In fact, the more I think about it, that just seems better.  In which case the question is, do we want it to default to leave one out?  Or just not have a default.  FWIW [`scikit-learn` has a default](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score) which (of course) is 3 folds ?!?.  I suppose that default is designed for dealing with really large datasets.